apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -0,0 +1,315 @@
+internal_resolve(Protocol, DNSMessage) ->
+    Response = erldns_handler:do_handle(DNSMessage, ?LOCALHOST),
+    encode_message(Protocol, Response).
+
+-spec(resolve(protocol(), [upstream()], binary(), reply_fun()) -> ok).
+resolve(Protocol, Upstreams, Data, Fun) ->
+    Workers =
+        lists:map(fun (Upstream) ->
+            Pid = start_worker(Protocol, Upstream, Data),
+            MonRef = monitor(process, Pid),
+            {Upstream, Pid, MonRef}
+        end, Upstreams),
+    resolve_loop(Workers, Fun).
+
+-spec(resolve_loop([{upstream(), pid(), reference()}], reply_fun()) -> ok).
+resolve_loop([], _Fun) ->
+    ok;
+resolve_loop(Workers, Fun) ->
+    receive
+        {reply, Pid, ReplyData} ->
+            {value, {Upstream, Pid, MonRef}, Workers0} =
+                lists:keytake(Pid, 2, Workers),
+            reply(Fun, ReplyData),
+            erlang:demonitor(MonRef, [flush]),
+            report_status([?MODULE, Upstream, successes]),
+            resolve_loop(Workers0, fun ok_reply_fun/1);
+        {'DOWN', MonRef, process, Pid, _Reason} ->
+            {value, {Upstream, Pid, MonRef}, Workers0} =
+                lists:keytake(Pid, 2, Workers),
+            report_status([?MODULE, Upstream, failures]),
+            resolve_loop(Workers0, Fun)
+    after 2 * ?TIMEOUT ->
+        lists:foreach(fun ({Upstream, Pid, MonRef}) ->
+            erlang:demonitor(MonRef, [flush]),
+            report_status([?MODULE, Upstream, failures]),
+            Pid ! {timeout, self()}
+        end, Workers)
+    end.
+
+%%%===================================================================
+%%% Workers
+%%%===================================================================
+
+-spec(start_worker(protocol(), upstream(), binary()) -> pid()).
+start_worker(Protocol, Upstream, Data) ->
+    Pid = self(),
+    proc_lib:spawn(fun () ->
+        init_worker(Protocol, Pid, Upstream, Data)
+    end).
+
+-spec(init_worker(protocol(), pid(), upstream(), binary()) -> ok).
+init_worker(udp, Pid, Upstreams, Data) ->
+    udp_worker(Pid, Upstreams, Data);
+init_worker(tcp, Pid, Upstreams, Data) ->
+    tcp_worker(Pid, Upstreams, Data).
+
+-spec(udp_worker(pid(), upstream(), binary()) -> ok).
+udp_worker(Pid, Upstream = {IP, Port}, Data) ->
+    StartTime = erlang:system_time(millisecond),
+    MonRef = erlang:monitor(process, Pid),
+    Opts = [{reuseaddr, true}, {active, once}, binary],
+    {ok, Socket} = gen_udp:open(0, Opts),
+    try
+        ok = gen_udp:send(Socket, IP, Port, Data),
+        receive
+            {'DOWN', MonRef, process, _Pid, normal} ->
+                ok;
+            {'DOWN', MonRef, process, _Pid, Reason} ->
+                exit(Reason);
+            {udp, Socket, IP, Port, ReplyData} ->
+                Pid ! {reply, self(), ReplyData},
+                report_latency([?MODULE, Upstream, latency], StartTime);
+            {timeout, Pid} ->
+                exit(timeout)
```

Comment:
would this appear in log? If not then probably we should explicit log this error